### ISSUE TITLE
Video binary object extraction

### ISSUE THREAD

**Issue by ruebot** [archivesunleashed/aut#306]
Using the [image extraction process](https://github.com/archivesunleashed/aut/pull/234) as a basis, our next set of binary object extractions will be documents. This issue is meant to focus specially on video objects. 

There may be a some tweaks to this depending on the outcome of #298.

**Comment by ruebot** [comment]
Ok, I have a basic framework setup in [the branch](https://github.com/archivesunleashed/aut/tree/issue-306-307).

Pull down the branch, and do something along these lines (I had memory issues, so I did my whole Spark config thing):

```
rm -rf ~/.m2/repository/* && mvn clean install && rm -rf ~/.ivy2/* && ~/bin/spark-2.4.3-bin-hadoop2.7/bin/spark-shell --master local\[10\] --driver-memory 35g --conf spark.network.timeout=10000000 --conf spark.executor.heartbeatInterval=600s --conf spark.driver.maxResultSize=4g --conf spark.serializer=org.apache.spark.serializer.KryoSerializer --conf spark.shuffle.compress=true --conf spark.rdd.compress=true --packages io.archivesunleashed:aut:0.17.1-SNAPSHOT -i ~/306-pdf-audio-video-extract.scala
```

**306-pdf-audio-video-extract.scala**
```scala
import io.archivesunleashed._
import io.archivesunleashed.df._

val df_pdf = RecordLoader.loadArchives("/home/nruest/Projects/au/sample-data/geocites/1/*gz", sc).extractPDFDetailsDF();
val res_pdf = df_pdf.select($"bytes", $"extension").saveToDisk("bytes", "/home/nruest/Projects/au/sample-data/306-307-test/pdf", "extension")

val df_audio = RecordLoader.loadArchives("/home/nruest/Projects/au/sample-data/geocites/1/*gz", sc).extractAudioDetailsDF();
val res_audio = df_audio.select($"bytes", $"extension").saveToDisk("bytes", "/home/nruest/Projects/au/sample-data/306-307-test/audio", "extension")

val df_video = RecordLoader.loadArchives("/home/nruest/Projects/au/sample-data/geocites/1/*gz", sc).extractVideoDetailsDF();
val res_video = df_video.select($"bytes", $"extension").saveToDisk("bytes", "/home/nruest/Projects/au/sample-data/306-307-test/video", "extension")

sys.exit
```

[I have a whole lot of audio, pdf, and video files](https://gist.github.com/ruebot/9287f260860a04269ab32a7bd088aa5d).

**Considerations**

1. We need tests. Should have had some for #302 :man_shrugging: 
2. Is this is ok implementation on getting the extension? It seems to have a really good success rate.
3. How do we want to handle when we don't or are not able to get an extension? Throw a conditional in the mix, and say if null/empty `UNKNOWN`? Do we want that stored in the dataframe, or done on the fly in `saveToDisk`?

@jrwiebe @lintool @ianmilligan1 let me know what you think.

**Comment by ianmilligan1** [comment]
Woohoo, this is looking great. Congrats @ruebot! I've tested locally on our [CPP Sample Data](https://github.com/archivesunleashed/aut-resources/tree/master/Sample-Data) and all the extractors are working on the data. Some fun PDFs and lots of weird political talk radio and interview clips. 

FWIW, at least on this weird CPP collection subset from 2009, it's having trouble getting any extensions for video (it found a few wmv and the rest are all sans extension). 

<img width="882" alt="Screen Shot 2019-08-12 at 5 41 57 PM" src="https://user-images.githubusercontent.com/3834704/62900570-7de4d100-bd28-11e9-94a8-50c4a0ca9b19.png">

I don't know the best route on your questions #2 and #3 so will leave that to the more qualified @jrwiebe and @lintool.

**Comment by jrwiebe** [comment]
I'm not sure about this method of getting the extension. I think the reason @ianmilligan1 was getting so many videos without extension is that you're just getting the extension based on the URL. It's easy to think of examples of how audio or video files might be served without containing the file extension. 

I think a better way to get the extension is based on the MIME type. There isn't a 1:1 mapping between MIME and extension, so perhaps we combine this with URL analysis. 

I've created [a branch that implements this method](https://github.com/archivesunleashed/aut/tree/get-extension). I just finished running [a test](https://gist.github.com/jrwiebe/fb09bea52b105289ba07140694c04275) with it right now. It was working fine until the end, when it [failed with `java.lang.OutOfMemoryError`](https://gist.github.com/jrwiebe/a7be3155caa34fc1e275a675b63d859e). So some thought resource use is warranted here.
***
(Aside: If you look at the branch's commit history you'll see I modified the POM. This is because the shading I did that somehow resolved #302 actually did not relocate (i.e., rename) `commons-compress` as intended. Evidently some other change allowed our tests to work, but I was getting that NoSuchMethodError related to commons-compress again when I tested my `getExtension` method. Now we're relocating for real, as verified by unzipping the JAR and seeing `shaded/org/apache/tika/tika-parsers/` paths.)
